witness/golang/internal/witness/witness_test.go

```diff
@@ -0,0 +1,329 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package witness
+
+import (
+	"context"
+	"crypto/rand"
+	"database/sql"
+	"encoding/hex"
+	"fmt"
+	"testing"
+
+	"github.com/google/trillian/merkle/logverifier"
+	"github.com/google/trillian/merkle/rfc6962/hasher"
+	_ "github.com/mattn/go-sqlite3" // Load drivers for sqlite3
+	"golang.org/x/mod/sumdb/note"
+)
+
+var (
+	initC = Chkpt{
+		Size: 5,
+		Raw:  []byte("Log Checkpoint v0\n5\n41smjBUiAU70EtKlT6lIOIYtRTYxYXsDB+XHfcvu/BE=\n"),
+	}
+	newC = Chkpt{
+		Size: 8,
+		Raw:  []byte("Log Checkpoint v0\n8\nV8K9aklZ4EPB+RMOk1/8VsJUdFZR77GDtZUQq84vSbo=\n"),
+	}
+	consProof = [][]byte{
+		dh("b9e1d62618f7fee8034e4c5010f727ab24d8e4705cb296c374bf2025a87a10d2", 32),
+		dh("aac66cd7a79ce4012d80762fe8eec3a77f22d1ca4145c3f4cee022e7efcd599d", 32),
+		dh("89d0f753f66a290c483b39cd5e9eafb12021293395fad3d4a2ad053cfbcfdc9e", 32),
+		dh("29e40bb79c966f4c6fe96aff6f30acfce5f3e8d84c02215175d6e018a5dee833", 32),
+	}
+)
+
+func signChkpts(skey string, chkpts []string) ([][]byte, error) {
+	ns, err := note.NewSigner(skey)
+	if err != nil {
+		return nil, fmt.Errorf("signChkpt: couldn't create signer")
+	}
+	signed := make([][]byte, len(chkpts))
+	for i, c := range chkpts {
+		s, err := note.Sign(&note.Note{Text: c}, ns)
+		if err != nil {
+			return nil, fmt.Errorf("signChkpt: couldn't sign note")
+		}
+		signed[i] = s
+	}
+	return signed, nil
+}
+
+func setOpts(d *Database, logID string, logPK string, wSK string) (*Opts, error) {
```

Comment:
`setOpts` -> `newOpts`? This function plays the role of `Opts` constructor. Same for other `set` functions below.

Reply:
Done.